lib/project_types/script/layers/infrastructure/script_uploader.rb

```diff
@@ -2,24 +2,46 @@ module Script
   module Layers
     module Infrastructure
       class ScriptUploader
+        CONTENT_LENGTH_RANGE_HEADER = "x-goog-content-length-range"
+
         def initialize(script_service)
           @script_service = script_service
         end
 
         def upload(script_content)
-          @script_service.generate_module_upload_url.tap do |url|
-            url = URI(url)
+          upload_details = @script_service.generate_module_upload_details
+          url = URI(upload_details[:url])
 
-            https = Net::HTTP.new(url.host, url.port)
-            https.use_ssl = true
+          https = Net::HTTP.new(url.host, url.port)
+          https.use_ssl = true
 
-            request = Net::HTTP::Put.new(url)
-            request["Content-Type"] = "application/wasm"
-            request.body = script_content
+          request = Net::HTTP::Put.new(url)
+          request["Content-Type"] = "application/wasm"
 
-            response = https.request(request)
-            raise Errors::ScriptUploadError unless response.code == "200"
+          headers = upload_details[:headers]
+          headers.each do |header, value|
+            request[header] = value
           end
+
+          request.body = script_content
+
+          response = https.request(request)
+          raise Errors::ScriptTooLargeError, file_size_limit(headers) if script_too_large?(response)
+          raise Errors::ScriptUploadError unless response.code == "200"
+
+          upload_details[:url]
+        end
+
+        private
+
+        def script_too_large?(response)
+          response.code == "400" && response.body.include?("EntityTooLarge")
+        end
+
+        def file_size_limit(headers)
+          content_length_range_value = headers[CONTENT_LENGTH_RANGE_HEADER]
+          return 0 unless content_length_range_value
+          content_length_range_value.split(",")[1].to_i
         end
```

Comment:
The fact that the CLI knows about this header and parses it out is kinda gross. WDYT about not sending arbitrary headers from Script Service, but instead actually sending the file size limit and having the CLI craft the request itself? On one hand, it's cleaner since we don't need to rely on specific headers being set, but on the other it does reduce our ability to change the headers if we do something similar with GCS in the future.